Sound trigger updates

ranj063 opened the pull request:
The main changes included are:
1. Unify the handling of all process type components.
2. Add support for KPB, channel selector and detect comp in the kernel
3. Add DAPM event handler that can trigger hostless pipelines upon receiving PRE_PMU/PRE_PMD DAPM events

plbossart commented on include/sound/sof/topology.h:
```diff
@@ -32,6 +32,9 @@ enum sof_comp_type {
+	SOF_COMP_KEYWORD_DETECT,
+	SOF_COMP_VOICE_DETECT,
+	/* keep FILEREAD/FILEWRITE as the last ones */
 	SOF_COMP_FILEREAD,	/**< host test based file IO */
```
you would want to set a value for SOF_COMP_FILEREAD, e.g. 10000 so that it "never" conflicts with regular component ever again?

plbossart commented on include/sound/sof/topology.h:
```diff
@@ -37,6 +37,8 @@ enum sof_comp_type {
 	SOF_COMP_FILEREAD,	/**< host test based file IO */
 	SOF_COMP_FILEWRITE,	/**< host test based file IO */
+	SOF_COMP_KPB, /* A key phrase buffer component */
+	SOF_COMP_SELECTOR, /* channel selector */
```
this isn't aligned with the note above to "keep FILEREAD/FILEWRITE as the last ones"

plbossart commented on include/sound/sof/topology.h:
```diff
@@ -173,6 +175,8 @@ enum sof_ipc_process_type {
 	SOF_PROCESS_KEYWORD_DETECT,	/**< Keyword Detection */
 	SOF_PROCESS_VOICE_DETECT,	/**< Voice Detection */
+	SOF_PROCESS_KPB,		/**< KeyPhrase Buffer Manager */
+	SOF_PROCESS_SELECTOR,		/**< Channel Selector */
```
selector is too vague.
SOF_PROCESS_CHAN_SELECTOR?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2714,6 +2746,10 @@ static int sof_route_load(struct snd_soc_component *scomp, int index,
 		goto err;
 	}
 
+	/* if source is of type "snd_soc_dapm_output" ignore it */
```
maybe explain why we ignore it for reviewers?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1898,6 +1898,12 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 		break;
+	/*
+	 * snd_soc_dapm_sink can treated as a special processing type widget.
+	 * So, re-use the same ipc structure and tokens as the effect widgets.
```
the explanation needs to be clarified. We could state the same thing for the siggen just above, why can't it be treated as an effect?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -38,11 +38,65 @@
+	pcm.comp_id = swidget->comp_id;
+	pcm.params.hdr.size = sizeof(pcm.params);
+
+	dev_dbg(sdev->dev, "ranjani: comp_id is %d\b", swidget->comp_id);
```
hello ranjani.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -52,7 +106,51 @@ static int sof_tplg_dapm_event(struct snd_soc_dapm_widget *w,
+				swidget->widget->name);
+
+		break;
+	case SND_SOC_DAPM_PRE_PMD:
```
POST_PMD? shouldn't this be symmetrical with PRE_PMU?

plbossart commented:
@mengdonglin @xiulipan does CI work on this branch?

keyonjie commented on include/sound/sof/topology.h:
```diff
@@ -32,8 +32,13 @@ enum sof_comp_type {
 	SOF_COMP_EQ_IIR,
 	SOF_COMP_EQ_FIR,
+	SOF_COMP_KEYWORD_DETECT,
+	SOF_COMP_VOICE_DETECT,
```
We actually not use voice_detect.

keyonjie commented on include/sound/sof/topology.h:
```diff
@@ -163,39 +168,26 @@ struct sof_ipc_comp_tone {
+	SOF_PROCESS_EQFIR,		/**< Intel FIR */
+	SOF_PROCESS_EQIIR,		/**< Intel IIR */
+	SOF_PROCESS_KEYWORD_DETECT,	/**< Keyword Detection */
+	SOF_PROCESS_VOICE_DETECT,	/**< Voice Detection */
```
shall we remove voice_detect and add it in future if we do need?

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -38,6 +38,125 @@
+	pcm.hdr.size = sizeof(pcm);
+	pcm.hdr.cmd = SOF_IPC_GLB_STREAM_MSG | SOF_IPC_STREAM_PCM_PARAMS;
+	pcm.comp_id = swidget->comp_id;
+	pcm.params.hdr.size = sizeof(pcm.params);
```
I am a little bit concern sending empty params here which means the detect pipeline won't support runtime format/rate change.

ranj063 replied:
@keyonjie lets bring it up in the meeting tomorrow

ranj063 replied:
fixed now to use runtime PCM params

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -38,6 +38,125 @@
+		break;
+	case SND_SOC_DAPM_PRE_PMD:
+		/* pcm free */
+		ret = ipc_trigger(swidget, SOF_IPC_STREAM_PCM_FREE);
```
trigger stop first then pcm free.

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -38,6 +38,125 @@
+				"error: failed to trigger widget %s\n",
+				swidget->widget->name);
+
+		break;
```
In my the integration test, I never get this pre/post_pmd event, we need to root cause it and verify the point that this event happen is suitable for trigger stop and free(some concern about the sequence and pipeline status).

ranj063 replied:
@keyonjie with out_drv type of virtual widgets i'm able to get the pre/post_pmd event. Please give it a try

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -235,26 +354,30 @@ static enum sof_ipc_frame find_format(const char *name)
+	{"EQFIR", SOF_PROCESS_EQFIR},
+	{"EQIIR", SOF_PROCESS_EQIIR},
+	{"KEYWORD_DETECT", SOF_PROCESS_KEYWORD_DETECT},
+	{"VOICE_DETECT", SOF_PROCESS_VOICE_DETECT},
```
remove voice detect.

keyonjie commented on sound/soc/sof/topology.c:
```diff
@@ -1865,8 +1996,15 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
+	 * So, re-use the same ipc structure and tokens as the effect widgets.
+	 */
+	case snd_soc_dapm_sink:
+		/* fallthrough */
```
Could this be conflict with the possible future use of the normal dapm_sink widget?

ranj063 commented:
@plbossart @keyonjie Ive addressed all your comments now. Main changes I made are:

1. use CHAN_SELECTOR instead of SELECTOR
2. stuck with effect type for keyword_detect. snd_soc_dapm_sink is an always on type component. So we cant use it for our needs.
3. removed VOICE_DETECT
4. check if keyword_detect comp has exactly 1 kcontrol 
5. check if the detect switch is on before activating the pipeline

ranj063 commented:
@plbossart @keyonjie removed the switch kcontrol requirement for the detect comp now

plbossart commented on include/sound/sof/topology.h:
```diff
@@ -33,6 +33,8 @@ enum sof_comp_type {
 	SOF_COMP_EQ_FIR,
 	SOF_COMP_KEYWORD_DETECT,
+	SOF_COMP_KPB, /* A key phrase buffer component */
+	SOF_COMP_SELECTOR,	/**< channel selector component */
```
consistency between comments?

ranj063 replied:
fixed

plbossart commented on include/sound/sof/topology.h:
```diff
@@ -33,6 +33,8 @@ enum sof_comp_type {
 	SOF_COMP_EQ_FIR,
 	SOF_COMP_KEYWORD_DETECT,
+	SOF_COMP_KPB, /* A key phrase buffer component */
+	SOF_COMP_SELECTOR,	/**< channel selector component */
```
SOF_COMP_CH_SELECTOR?

ranj063 replied:
@plbossart this will have to change in the FW as well. Let me submit a PR for the FW too

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -120,7 +146,7 @@ static int sof_tplg_dapm_event(struct snd_soc_dapm_widget *w,
 	switch (event) {
 	case SND_SOC_DAPM_PRE_PMU:
-		/* set pcm params. sink widgets can only do capture */
+		/* get pcm params */
```
Get or set?

ranj063 replied:
fixed :)

keyonjie commented:
Verified that Ranjani’s PR do send hw_params to selector component.
selector.c:187        frame_fmt: 0, periods_sink: 2

ranj063 commented:
> Verified that Ranjani’s PR do send hw_params to selector component.
> selector.c:187 frame_fmt: 0, periods_sink: 2

@keyonjie this only worked because of the bug in snd_sof_find_spcm_name(). I've fixed it now. Could you please try it again. If you do not set stream_name for the detect comp, this should fail now

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -243,6 +243,7 @@ struct sof_process_types {
 static const struct sof_process_types sof_processs[] = {
 	{"EQFIR", SOF_PROCESS_EQFIR},
 	{"EQIIR", SOF_PROCESS_EQIIR},
+	{"KEYWORD_DETECT", SOF_PROCESS_KEYWORD_DETECT},
```
maybe clarify why KPB and CHAN_SELECTOR are omitted here?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -243,6 +243,7 @@ struct sof_process_types {
 static const struct sof_process_types sof_processs[] = {
```
processsss? 2 s are ok

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1763,6 +1765,14 @@ static int sof_widget_load_process(struct snd_soc_component *scomp, int index,
+	case SOF_PROCESS_CHAN_SELECTOR:
+		ret = sof_process_load(scomp, index, swidget, tw, r,
+				       SOF_COMP_SELECTOR);
+		break;
```
this is starting to get ugly with most of the code copy/pasted, could we not use a table instead of a switch?

ranj063 replied:
@plbossart yes will do. I think this is what @juimonen was talking about to.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -52,7 +105,57 @@ static int sof_tplg_dapm_event(struct snd_soc_dapm_widget *w,
-	return 0;
+	/* nothing to do if widget is not of type effect */
+	if (swidget->id != snd_soc_dapm_effect)
+		return 0;
```
could we instead register this callback only for the KEYWORD_DETECT widget?

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1900,6 +1921,18 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
+						     tw->event_type);
+		if (ret) {
+			dev_err(sdev->dev, "error: widget event binding failed\n");
+			return -EINVAL;
```
as described above, maybe we can specialize this widget ops table and use for the detector, instead of having a generic table that only works for the Detector?

ranj063 replied:
sounds good. I will change this

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -29,13 +29,18 @@ struct snd_sof_pcm *snd_sof_find_spcm_name(struct snd_sof_dev *sdev,
 
-		if (strcmp(spcm->pcm.caps[0].name, name) == 0)
+		/* match with playback caps name if set */
+		if (strcmp(spcm->pcm.caps[0].name, "") &&
```
if (*str) ?

ranj063 replied:
@plbossart *str will be the null terminator. I think its better to check "if (strlen(spcm->pcm.caps[0].name &&" instead.

ranj063 replied:
oh I was wrong. just realized the null terminator has a value of 0. Thanks!

ranj063 commented:
@plbossart all comments addressed now

keyonjie commented on sound/soc/sof/core.c:
```diff
@@ -24,18 +24,23 @@
 
-		if (strcmp(spcm->pcm.caps[1].name, name) == 0)
+		/* match with capture caps name if set */
+		if (*spcm->pcm.caps[0].name &&
```
typo here which will make it always fail for capture-only PCM, should be caps[1].name

ranj063 replied:
@keyonjie thanks. fixed now

xiulipan commented:
@plbossart 
Seems we did not track branches other than sof-dev. I will check what should be done to enable more branch track.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -1781,6 +1924,39 @@ static int sof_widget_load_process(struct snd_soc_component *scomp, int index,
+	return snd_soc_tplg_widget_bind_event(swidget->widget,
+					      sof_tplg_widget_ops,
+					      ARRAY_SIZE(sof_tplg_widget_ops),
+					      event_type);
```
you can simplify the flow by moving this bind_event() call inside of the switch, and renaming the widget_ops as keyword_widget_ops.

ranj063 replied:
@plbossart updated now.

ranj063 commented:
@plbossart there's one style issue with a line exceeding 80 chars in the sof_widget_bind_event(). I think it should be OK to leave it as is. Please let me know if you feel otherwise

keyonjie commented:
@plbossart I think this PR is ready now for KWD integration testing, agree?

keyonjie commented:
merging it now.
